R/allocate_reference_times.R

```diff
@@ -0,0 +1,337 @@
+#'      Default is `3`.
+#' @param prop_delay Numeric value <1 indicating what proportion of all
+#'     reference times in the reporting triangle to be used for delay
+#'     estimation. Default is `0.5`.
+#' @param size_min_retro_nowcasts Integer indicating the minimum number of
+#'     reference times needed for uncertainty estimation. Default is `2`.
+#' @importFrom cli cli_abort cli_warn
+#' @importFrom checkmate assert_integerish
+#' @export
+#' @returns list of n_history_delay and n_retrospective_nowcasts
+#' @examples
+#' triangle <- matrix(
+#'   c(
+#'     100, 50, 30, 20,
+#'     40, 10, 20, 5,
+#'     80, 50, 25, 10,
+#'     100, 50, 30, 20,
+#'     40, 10, 20, 5,
+#'     80, 50, 25, 10,
+#'     100, 50, 30, 20,
+#'     90, 45, 25, NA,
+#'     80, 40, NA, NA,
+#'     70, NA, NA, NA
+#'   ),
+#'   nrow = 10,
+#'   byrow = TRUE
+#' )
+#' # Use the defaults
+#' ref_time_allocation_default <- allocate_reference_times(triangle)
+#' ref_time_allocation_default
+#' # Modify to use less volume and redistribute
+#' ref_time_allocation_alt <- allocate_reference_times(
+#'   reporting_triangle = triangle,
+#'   scale_factor = 2,
+#'   prop_delay = 0.6
+#' )
+#' ref_time_allocation_alt
+allocate_reference_times <- function(reporting_triangle,
+                                     max_delay = ncol(reporting_triangle) - 1,
+                                     scale_factor = 3,
+                                     prop_delay = 0.5,
+                                     size_min_retro_nowcasts = 2) {
+  # Checks of inputs
+  .validate_triangle(reporting_triangle, max_delay)
+  .validate_allocation_params(
+    scale_factor, prop_delay,
+    size_min_retro_nowcasts
+  )
+
+  ns <- .perform_allocation_process(
+    reporting_triangle,
+    max_delay,
+    scale_factor,
+    prop_delay,
+    size_min_retro_nowcasts
+  )
+
+  return(ns)
+}
+
+#' Perform the allocation process
+#'
+#' @inheritParams allocate_reference_times
+#'
+#' @return list of reference time allocations
+.perform_allocation_process <- function(reporting_triangle,
+                                        max_delay,
+                                        scale_factor,
+                                        prop_delay,
+                                        size_min_retro_nowcasts) {
+  sizes <- .calculate_sizes(
+    reporting_triangle,
+    max_delay,
+    scale_factor,
+    size_min_retro_nowcasts
+  )
+  n_ref_times <- sizes$n_ref_times
+  size_required <- sizes$size_required
+  size_min_delay <- sizes$size_min_delay
+  size_target <- sizes$size_target
+
+  # Check specifications against targets and requirements
+  size_used <- .check_against_requirements(
+    n_ref_times,
+    size_required,
+    size_target,
+    size_min_delay,
+    size_min_retro_nowcasts,
+    scale_factor,
+    max_delay
+  )
+
+  # If size being used equals or exceeds the target, simply split according to prop delay #nolint
+  ns <- .assign_ns_from_sizes(
```

Comment:
i.e here if we didn't say size at all possible opporuntities we would have text that was 30% of the length